syntax = "proto3";

import "google/protobuf/timestamp.proto";

option csharp_namespace = "GrpcCloud";


enum Status {
    SUCCESS = 0;
    FAILURE = 1;
}

message FileMetadata{
	string filename = 1;
	string type = 2;
	string creatorEmail = 3;
	int32 size = 4;
	google.protobuf.Timestamp creationDate = 5;
    google.protobuf.Timestamp  lastModified = 6;
}

message loginRequest {
	string username = 1;
    string password = 2;
}

message loginResponse {
	Status status = 1;
	string sessionId = 2;
}

message signupRequest {
	string username = 1;
    string password = 2;
	string email = 3;
	string phoneNumber = 4;
}

message signupResponse {
  Status status = 1;
  string message = 2;
}

message logoutRequest {
	string sessionId = 1;
}

message logoutResponse {

}

message UploadFileRequest {
	string sessionId = 1;
    string fileName = 2;
    string type = 3;
    bytes fileData = 4;
}

message UploadFileResponse {
	Status status = 1;
	string message = 2;
}

message DownloadFileRequest {
	string sessionId = 1;
	string fileName = 2;
}

message DownloadFileResponse {
	Status status = 1;
	string message = 2;
	bytes fileData = 3;
}

message DeleteFileRequest {
	string sessionId = 1;
    string fileName = 2;
}

message DeleteFileResponse {
	Status status = 1;
	string message = 2;
}

message getListOfFilesRequest{
	string sessionId = 1; 
}

message getListOfFilesResponse{
	Status status = 1;
	string message = 2;
	repeated FileMetadata files = 3;
}

message getFileMetadataRequest{
	string sessionId = 1;
    string fileName = 2;
}

message getFileMetadataResponse{
	Status status = 1;
	string message = 2;
	FileMetadata file = 3;
}

service Cloud {

	// Auth:
	rpc login(loginRequest) returns (loginResponse);
	rpc signup(signupRequest) returns (signupResponse);
	rpc logout(logoutRequest) returns (logoutResponse);
	
	// Metadata:
	rpc getListOfFiles(getListOfFilesRequest) returns(getListOfFilesResponse);
	rpc getFileMetadata(getFileMetadataRequest) returns(getFileMetadataResponse) ;

	// File methods:
	rpc UploadFile(stream UploadFileRequest) returns (UploadFileResponse);
    rpc DownloadFile(DownloadFileRequest) returns (stream DownloadFileResponse);
    rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse);
}
